8.0 add modules intrastat_base and intrastat_product (v3)

alexis-via opened the pull request:
My third attempt at migrating intrastat_base and intrastat_product from lp:new-report-intrastat to github/OCA

Commit:
Initial commit of the entirely new code for report_intrastat for France.

Commit:
Better performances : we now use multi= to compute numbers of lines and total amount in a single function call.

Commit:
Add currency in view via a fields.related on report_intrastat_service
Check that company currency is EUR when generating service lines.
Do the rounding work when generating the service lines, not when generating the XML.
Remove unused code.

Commit:
View style update

Commit:
Started work on l10n_fr_intrastat_product

Commit:
First alpha version of the new module "l10n_fr_intrastat_product".
First factorisation of common functionnal code between the product and service module (intrastat_common.py) - to be continued.
Some code clean-up in the service module.

Commit:
More work on full DEB+DES implementation.
Added full support for "transport type" and partial support for the department.
More factorisation between product & service module.
Add date_done field.
More checks.

Commit:
Converted some fields on report_intrastat_product/service to many2one (fields that are not in the official declaration or never subject to change)
Add copy of department from warehouse to invoice.
Import DEB don't have the partner_vat field
Department can be 99 for Monaco.
Copy tranport and department from report_intrastat_product in XML file
First attempt at adapting the display of fields depending on the type of declaration.

Commit:
In report_intrastat_product_line, remplace the 2 country fields from string to many2one.

WARNING : this is my last commit before I break everything to re-organise the way intrastat_types are handled and I drop the "Intrastat type" field on invoices.

Commit:
Complete change of design for intrastat_type !
Now that report_intrastat_product/service object is stored in database, we don't need the legal_intrastat state on invoices any more, because we will be able to create entry lines for DEB for a repair of equipment under warranty without using a legal_intrastat.
Report_intrastat_type is now only required for DEB -> object moved to module l10n_fr_intrastat_product.
For DEB, the button "regenerate lines" only regenerate DEB lines related to invoices.

Renamed intrastat base module, because it doesn't have France-specific parameters any more.

Add demo data.

Commit:
Renamed some fields
Now manages the specifications of the different procedure_code, for example some info don't need to be declared for some procedure codes.
Ability to enter DEB lines manually
Stat value doesn't exist any more
Tree vue of DEB lines is now identical to Cerfa

Commit:
"Date done" converted to date + time.
Re-order fields in intrastat_service view

Commit:
Better interface when entering DEB lines manually.
Commit before I break the main view of intrastat_product.

Commit:
Brand new XML view for intrastat_product, for better usability (dynamic readonly & required)

Commit:
[FIX] Move the check on weight higher in the code to avoid a crash when weight = None

Commit:
Removed from prints (not all yet)

Commit:
[FIX] Error in the inherit of view, which can cause a crash in the product view.
Beta implementation of "Re-generate lines from picking".
[FIX] Check on weight_net instead of weight.

Commit:
Desactivate a check temporaryly.

Commit:
Add pricelist on res_company to compute the stat value
Ability to factorize lines taken from picking
renamed field invoice_uom_id to source_uom_id, because this field is now also used for lines taken from picking
Removed field date_invoice
Removed the label "beta" on the button "generate lines from picking".

Commit:
- demo data : Add department on warehouse
- removed the use of "user" in the function compute_lines_from_picking
- add a field.function "intrastat department" on picking
- updated the code accordingly : we can now have the department of the picking when generating lines from picking.

Commit:
Removed unused code in SQL query (this "id" field was required when it was an SQL view)

Commit:
- For "generate lines from picking", exclude picking from sale orders/purchase orders that have a shipping policy != 'invoice from the packing'
So it should now only select incoming/outgoing products used for RMAs

Commit:
Full re-design of intrastat types : probably requires deleting the report_intrastat_type table, restarting OpenERP and re-creating intrastat types.
Moved intrastat departments from stock.warehouse to stock.location.
Dropped SQL queries ; replaced by traditionnal python code logic. No more need to have one rate per day for invoices with foreign currency.
Add total fiscal value for DEB
More code factorization.
Prepare translation work.

Commit:
New important feature : ability to exlude invoice lines that have a specific tax in it.
For example, this is required for B2C companies, that mustn't put in DEB Export the invoices to EU customers that contain French VAT.

Commit:
Label updated.

Commit:
[FIX] Bug when opening picking.

Commit:
Support for some triangular operations (not all yet)
[FIX] an issue with the type of field of transaction_code on intrastat_type
Add field "Destination/Origin country of the goods" on invoice
Write the value of this field when creating the invoice from the packing
Fix a bug in the computation of the department on picking, and now make it a stored field.
Now able to copy origin/dest country and department on invoice when the invoice is created from SO/PO (before, it was only supported when creating the invoice from the packing).

Commit:
Moved some demo data from l10n_fr_intrastat_product to intrastat_base
Moved configuration about taxes from company form to tax form
Some modifications to ease v5 -> v6 migration :
- object report_intrastat_code now belong to group account manager
- button functions now return True
Tried to implement the following feature : open attachement form when the XML file as been generated : works on v6, but make client crash en v5 -> code has been commented
DEB lines with procedure code = 25 are now deducted from the fiscal total.
Round invoice total.

Commit:
Moved the field exclude_from_intrastat_if_present of account.tax from l10n_fr_intrastat_product to intrastat_base, because it should also be used in the module l10n_fr_intrastat_service.
Take this field into account in the generation of DEB lines (module l10n_fr_intrastat_service).

Commit:
User interface enhancements.

Commit:
Remove check that would block the manual creation of intrastat product lines.
Auto-fill the intrastat UoM field when creating an intrastat product line manually.

Commit:
More fixes in manual creation of DEB lines (more to come...)

Commit:
[FIX] partner VAT is now saved when entering the DEB line manually.

Commit:
It is now possible to set the H.S. code on the product category.
Some "return None" changed to "return True"

Commit:
. Implemented the fiscal representative : for example, when you ship to the EU but invoice outside of the EU, your customer needs to have a fiscal representative inside the EU, which will be used for the DEB
. depend on base_vat instead of account.

Commit:
Update of some strings before translation.
Started translation work.

Commit:
Better error message when fiscal rep is missing.

Commit:
Some fixes in strings.
Finished translation to French.

Commit:
Add missing import

Commit:
Modify demo data.

Commit:
Translation and strings update.
Add demo data.

Commit:
Port to OpenERP v6
Add seach views for Intrastat product and Intrastat service
Comment prints in the code
Standardize the name of ids
Add "done" button in tree view
Display company_id fields in the interface if the user belongs to "group_multi_company"

Commit:
Remove select="2", which is useless in v6

Commit:
l10n_fr_intrastat_service : Moved the raise if the Partner VAT number is missing, so that, when we sell to a physical person in the EU, we don't get the raise.
Thanks to Anevia for this interesting bug report !
We need to do the same thing in l10n_fr_intrastat_product.

Commit:
Same modification as my previous commit for DEB : when we sell to a physical person in the EU with VAT, the move is not declared in DEB, so it must not block with a "raise" if the partner doesn't have a VAT number.

Commit:
Update translations.

Commit:
Updated licence pointer.

Commit:
Copy-paste comment error fixed !

Commit:
Now manage UoM other than PCE. If the UoM required by the H.S. code is not the UoM of the invoi
ce, it will do the conversion (if possible).
Now possible to have a country of origin on the 'supplier information' instead of on the product form (usefull when the origin of a product depend on the supplier).

Commit:
Modify help message.

Commit:
Longer description for HS code.

Commit:
Now support weight computation when UoM category is "Unit" and UoM != PCE
Factorize code

Commit:
Fixes bug 955065 : don't declare DEB lines with value = 0

yvaucher committed:
[FIX] l10n_fr_intrastat_product - SaleOrder's method action_invoice_create signature

Commit:
[Merge] with lp:~c2c/new-report-intrastat/l10n_fr_intrastat_product-fix-sale-20120510-2
Thanks to Yannick Vaucher from Camptocamp for this fix.

Commit:
Add option "is_accessory_cost" on product.template :
  If the invoice has is_accessory_cost services but no regular product -> DES
  If the invoice has is_accessory_cost services and regular product -> added to the cost of products in DEB

Now allows "pricelist for statistical value" which is not in EUR (the currency conversion will be made from the pricelist currency to EUR)

Usability improvements :
- Order for DEB and DES tree view : "the more recent at the top"
- distinction between "Information to declare" and "Additionnal information" in intrastat lines

Commit:
Better string.
Update French translation.

Commit:
Update FR translation.

Commit:
Add demo value for department.

Commit:
Do not inherit action_invoice_create in sale.py and stock.py
We now inherit _prepare_invoice... thanks to Akretion's contribution in OpenERP 6.1 !

Commit:
Use the new logger API of OpenERP 6.1
Move the "EU fiscal representative" field in order to avoid the "compression" of the VAT field that made it too small.

Commit:
default -> defaults

Commit:
Rewrote the constraint of product.supplierinfo, for better perfs.
Thanks to Benoit Guillot for pointing out this perf issue.

Commit:
Code clean-up :
- context is not passed in constraints
- don't use lambda when not necessary

Commit:
[ADD] Module for manual declaration of Dutch intrastat report

Commit:
[ADD] model permissions
[FIX] class redefinition

Commit:
[IMP] Prevent lines with only zero amounts

Commit:
[DEL] Old TODO

Commit:
[FIX] Cannot delete a draft report

Commit:
[FIX] Set intrastat option for France in NL module

Commit:
[FIX] Allow for invoice lines without a product

Commit:
IMPORTANT CHANGE :
- All EU countries should now be intrastat=True, including your own country
- When generating lines for Intrastat Product/Service, all invoices for which country == Company's country are excluded

Commit:
As stated in BOD dated 05/01/2012, mouvements for repairs are not required any more in DEB since 01/01/2012.
Button has been removed. Lines of code will be removed later.

Commit:
Update help message according to my change of commit 61.
Fix copyright header.

Commit:
Initial port to v7.0. Not tested a lot for the moment.

Commit:
Continue removal of mouvements for repairs (cf BOD dated 05/01/2012)

Commit:
[MRG] Merged with lp:new-report-intrastat rev. 63

Commit:
[FIX] Remove obsolete reconfiguration of France's intrastat setting
	that was not even loaded in the first place (ouch!)

Commit:
[RFR] Style and conventions

Commit:
Fix wrong code.

Commit:
Introduce a big hack to make the fields added on stock.picking work
Remove method=True
Add new tab "Intrastat" on invoice views, with all intrastat-related fields
Code cleanup

Commit:
Add report_eur1 module, that proposes the official EUR1 report (requires report_aeroo).

Commit:
Fake commit.

Commit:
Adapt views to new v7.0 style.
Replace "note" field by the mail.thread object.

Commit:
Handle the case where we have in an invoice :
- some HW products with value = 0
- some accessory costs
=> we want to have the accessory costs in DEB, not in DES (before this fix, the accessory costs were in both)

Commit:
Handle the case where an invoice has products with 0 value and accessory costs with not null value -> the accessory costs are affected at the pro-rata of the number of lines.

Commit:
Use the GREAT feature widget="monetary"

Commit:
[FIX] on picking, address_id -> partner_id

Commit:
[ENH] Added Dutch Intrastat report, courtesy of Stefan Rijnhart (Therp BV).

Commit:
Should now be compatible with both OCB and RS-OCB
Add field import_obligation_level on res.company
Add group group_detailed_intrastat_product, so that companies that only use obligation = simplified don't see all the additionnal fields.
Remove transaction code corresponding to repairs in intrastat types
Better on_change on intrastat types (code is mutualised with field.function)
Update syntax : demo_xml/update_xml/init_xml -> data/demo

Commit:
[FIX] Crash on intrastat_product.py line 334 reported by Frederic Clementi.

Commit:
Welcome to Croatia in the European Union !

Commit:
[UPG] Migration to 7.0

Commit:
Add an e-mail reminder for l10n_fr_intrastat_product and l10n_fr_intrastat_service (hope that Akretion France won't forget it's own declarations now !!!)
On report.intrastat.product and report.intrastat.service : add copy() fonctions, tracking of important fields, a year_month function field and enhance views. Remove date_done field (the tracking in the chatter does the job).
Remove class instanciation in the code.

Commit:
[IMP] Better form view of product category, courtesy of David Beal.

Commit:
Raise an explicit exception (take into account Stefan remark on the merge proposal)

Commit:
[MERGE] with branch 70-add-reminder
Add an e-mail reminder for l10n_fr_intrastat_product and l10n_fr_intrastat_service (hope that Akretion France won't forget it's own declarations now !!!)
On report.intrastat.product and report.intrastat.service : add copy() fonctions, tracking of important fields, a year_month function field and enhance views. Remove date_done field (the tracking in the chatter does the job).
Remove class instanciation in the code.
Better form view of product category, courtesy of David Beal.

Commit:
[MERGE] branch 7.0-UPG-l10n_nl_intrastat
Contribution by Stefan Rijnhart (Therp)
Migration of module l10n_nl_intrastat to OpenERP 7.0

Commit:
[usability] Add the H.S. code description in the display name of the H.S. code

Commit:
Fix to make the module truly usable when user is not part of the group "Detailed intrastat product".
Remove dead code and fields that was used when we had to put DEB lines for repair operations (a thing of the past !).
Update coding style.
Reduce the number of flake8 warnings.

Commit:
[FIX] On OpenERP 7, when you have the document module installed and you download the attachement via the drop down list on the form view, the name of the file will be the name of the attachement and not datas_fname ; so we need to have name = datas_fname.

Commit:
Remove irrelevant comment.

Commit:
Add ir.rule for multi-company.
Better "OpenERP v7 style/ergonomy" for FR intrastat reports

Commit:
Only send reminder for French companies.

bealdav committed:
[FIX] switch required=True for country_id from model to views in res.partner

bealdav committed:
[IMP] xpath in partner view as Niels Huylebroeck suggest

bealdav committed:
[IMP] simplify first xpath too in partner view

bealdav committed:
[IMP] xpath in a more generic way

Commit:
[Merge] with branch lp:~akretion-team/new-report-intrastat/70-new-intrastat-switch-required-country_id-to-view-dbl with minor modifications.

Commit:
Use l10n_fr_* : use the method _company_default_get()
intrastat_base :
- the module is now PEP-8 compliant !
- add search view on countries
- is_accessory_cost is now invisible when product != service
- update POT file and FR translation
- remove double import of logging lib

Commit:
Add context in on_change (to be able to use web_context_tunnel)
XSD files are now pure XSD files, not python file with the content as string.
Convert l10n_fr_intrastat_service to PEP-8
Start to convert l10n_fr_intrastat_product to PEP-8 (not finished yet)
l10n_fr_intrastat_product : Update POT file and FR translation.

Commit:
Set fields country_id and intrastat_id invisible when type = service.

Commit:
l10n_fr_intrastat_service : update POT file and FR translation.
l10n_fr_intrastat_product : continue PEP-8 compliance

Commit:
Port intrastat_base to Odoo 8.0.

Commit:
Continue the work on port to v8.0

Commit:
Add intrastat_type_data and update demo data accordingly
Remove version="7.0" from form views
Add ondelete='restrict' on M2O pointing to intrastat.type
Add graph views on intrastat.product and intrastat.service... because Odoo v8 graph views are so cool ! :)

Commit:
Update graph view for intrastat.product, to split between import and export by default
Add dependancy on l10n_fr_siret (works on v8 without changes) and remove field siret_complement
Update demo data
PEP8 stuff

Commit:
Inherit the new _prepare method that has been introduced in v8.0 for generating invoices from POs.

Commit:
[IMP] When the obligation level for import is none, the type of the DEB is automatically set to export.
[FIX] country -> country_id
type='string' -> type='char' (v8 only accepts char)
Change "error msg in raise", because v8 displays this to the user
Remove statistical_pricelist_id from demo data
Harmonize labels of button between DEB and DES

Commit:
Add module intrastat_product.
Rename report.intrastat.product to l10n.fr.report.intrastat.product (same for service and for lines) + migration scripts

Commit:
Rename field country_id to origin_country_id (TODO : write mig script) and move this field from l10n_fr_intrastat_product to intrastat_product

Commit:
Take into accout the remarks of Guewen

Commit:
Move intrastat_code and uom_id of report.intrastat.code from l10n_fr_intrastat_product to intrastat_product, as discussed with Luc de Meyer

Commit:
Merge form lp:~akretion-team/new-report-intrastat/8.0-add-intrastat-product

Commit:
[FIX] Forgot rename of model in service reminder file

Commit:
intrastat_base + l10n_fr_intrastat_service : port to new API

Commit:
Continue the move to the new API/flake8

Commit:
Port intrastat.type to the new API

NL66278 committed:
[FIX] Make dutch ICP report workable on odoo 8.0
[FIX] Some small changes in base, to make check_generate_lines better
    reusable (no need to force currency field on report), and do away with
    '_' before method name, as method should not be private, as it is called
    from other classes.

NL66278 committed:
[FIX] Use new API to prevent pylint errors and streamline code.

Commit:
Continue the port to the new API
Access for intrastat report to accounting user and not only accounting manager.

Commit:
Continue port to new API + PEP8. First working version.

Commit:
FIX validation of department

NL66278 committed:
[FIX] Some style improvements.

    Courtesy of comments by Alexis de Lattre.

NL66278 committed:
[REVERT] Revert rename of _check_generate_lines to version withouth underscore.

    In pending merge request from Alexis, _check_generate_lines will no
    longer be called from outside the class, so no longer any need to change
    private function name convention to public name convention.

    This will also make it easier to merge all changes.

Commit:
[MERGE] from lp:~therp-nl/new-report-intrastat/8.0-update-l10n_nl_intrastat
l10n_nl_intrastat : Port to new API + PEP8

Commit:
MERGE with trunk

Commit:
Use display_name instead of an inherit of name_get (better use of new API)
Inherit write and create of intrastat codes to remove spaces (patch by Luc De Meyer from Noviat)
Add O2M fields from intrastat code to products and product categs

Commit:
Move description from __openerp__.py to README.rst
Add conflict key in __openerp__.py
Simplify code that create attachment
Thanks to Luc de Meyer (Noviat) for his help on this.

Commit:
Fix a crash in Import DEB
Update strings

Commit:
Update translation and adapt a few strings

Commit:
PEP8 compliant
Port a line of code that was commented with a TODO

Commit:
report_eur1: port to v8

Commit:
Remove comment

Commit:
MERGE with Akretion's branch '8.0-fr-intrastat-new-api' : port to new API with a few minor enhancements

Commit:
l10n_fr_intrastat_service: update string and translation

Commit:
display_name must be stored as it is used as _rec_name, otherwise we cant, for instance, import files with name_search

Commit:
MERGE lp:~camptocamp/new-report-intrastat/store-name-search
display_name must be stored

Commit:
Merge lp:new-report-intrastat into 8.0-add-intrastat-v3

Commit:
Try to shut-up pylint

yvaucher commented on intrastat_base/README.rst:
```diff
@@ -0,0 +1,29 @@
+
+Author
+------
+* Alexis de Lattre, Akretion <alexis.delattre@akretion.com>
```
Missing maintainer

yvaucher commented on intrastat_base/__openerp__.py:
```diff
@@ -0,0 +1,44 @@
+    'category': 'Localisation/Report Intrastat',
+    'license': 'AGPL-3',
+    'summary': 'Base module for Intrastat reporting',
+    'author': 'Akretion',
```
Missing OCA in authors

yvaucher commented on intrastat_base/i18n/fr.po:
```diff
@@ -0,0 +1,191 @@
+#. module: intrastat_base
+#: constraint:product.template:0
+msgid "Error msg is in raise"
+msgstr "Error msg is in raise"
```
Not translated.

Keep it blank or translate it.

yvaucher commented on intrastat_base/i18n/fr.po:
```diff
@@ -0,0 +1,191 @@
+
+#. module: intrastat_base
+#: help:res.company,intrastat_remind_user_ids:0
+msgid "List of OpenERP users who will receive a notification to remind them about the Intrastat declaration."
```
s/OpenERP/Odoo

yvaucher commented on intrastat_base/i18n/fr.po:
```diff
@@ -0,0 +1,191 @@
+#. module: intrastat_base
+#: help:res.company,intrastat_remind_user_ids:0
+msgid "List of OpenERP users who will receive a notification to remind them about the Intrastat declaration."
+msgstr "Liste d'utilisateurs OpenERP qui recevront le rappel pour la DEB et/ou la DES."
```
s/OpenERP/Odoo

yvaucher commented on intrastat_base/i18n/intrastat_base.pot:
```diff
@@ -0,0 +1,191 @@
+
+#. module: intrastat_base
+#: help:res.company,intrastat_remind_user_ids:0
+msgid "List of OpenERP users who will receive a notification to remind them about the Intrastat declaration."
```
s/OpenERP/Odoo

yvaucher commented on intrastat_base/company.py:
```diff
@@ -0,0 +1,56 @@
+    intrastat_remind_user_ids = fields.Many2many(
+        'res.users', column1='company_id', column2='user_id',
+        string="Users Receiving the Intrastat Reminder",
+        help="List of OpenERP users who will receive a notification to "
```
s/OpenERP/Odoo

yvaucher commented on intrastat_base/i18n/fr.po:
```diff
@@ -0,0 +1,191 @@
+#. module: intrastat_base
+#: constraint:res.company:0
+msgid "error msg in raise"
+msgstr "error msg in raise"
```
Not translated.

To remove?

yvaucher commented on intrastat_base/i18n/intrastat_base.pot:
```diff
@@ -0,0 +1,191 @@
+
+#. module: intrastat_base
+#: constraint:res.company:0
+msgid "error msg in raise"
```
To remove? Does this translation come from debug?

yvaucher commented on intrastat_base/i18n/intrastat_base.pot:
```diff
@@ -0,0 +1,191 @@
+
+#. module: intrastat_base
+#: constraint:product.template:0
+msgid "Error msg is in raise"
```
To remove? Does this translation come from debug?

yvaucher commented on intrastat_base/intrastat_common.py:
```diff
@@ -0,0 +1,167 @@
+
+    @api.one
+    @api.depends(
+        'intrastat_line_ids', 'intrastat_line_ids.amount_company_currency')
```
You can depends only on `intrastat_line_ids.amount_company_currency` as it implies `intrastat_line_ids`

yvaucher commented on intrastat_base/intrastat_common.py:
```diff
@@ -0,0 +1,167 @@
+    def send_reminder_email(self, mail_template_xmlid):
+        mail_template = self.env.ref(mail_template_xmlid)
+        if self.company_id.intrastat_remind_user_ids:
+            self.pool['email.template'].send_mail(
```
I'm curious isn't it possible with API 8.0 to do as in event module:
https://github.com/odoo/odoo/blob/8.0/addons/event/event.py#L378

```
mail_template.send_mail(self.id)
```

If not, a comment would be welcome to explain why as this will certainly need to be rewritten in next Odoo version

yvaucher commented on intrastat_product/__openerp__.py:
```diff
@@ -0,0 +1,45 @@
+    'category': 'Localisation/Report Intrastat',
+    'license': 'AGPL-3',
+    'summary': 'Base module for Intrastat Product',
+    'description': """
```
Please put this in a README.rst

yvaucher commented on intrastat_product/__openerp__.py:
```diff
@@ -0,0 +1,45 @@
+This module has been written by Alexis de Lattre from Akretion
+<alexis.delattre@akretion.com>
+    """,
+    'author': 'Akretion',
```
OCA is missing

yvaucher commented on intrastat_product/intrastat.py:
```diff
@@ -0,0 +1,132 @@
+        self.display_name = display_name
+
+    name = fields.Char(
+        string='H.S. code',
```
You probably want an index on this?

`index=True`

yvaucher commented on intrastat_product/intrastat.py:
```diff
@@ -0,0 +1,132 @@
+    display_name = fields.Char(
+        compute='_compute_display_name', string="Display Name", readonly=True,
+        store=True)
+    intrastat_code = fields.Char(
```
And an index on this could help too

yvaucher commented:
@alexis-via See my inline comments, Plus it would be great with few small tests on added functions.

Commit:
Update README.rst and translation files

Commit:
Simplify @depends, as suggested by Yannick

Commit:
Better use of the new API for send_mail(), as suggested by Yannick (and tested be me)

Commit:
Add index=True on some fields on which we often search, as suggested by Yannick

alexis-via commented:
@yvaucher Thanks for your review ; I took into account all your remarks.
Sorry for the lack of update of README.rst and author field ; I did that in my (fucked) first PR, but i forgot to re-do it in my second (fucked) PR and third (good) PR !

brother-bernard commented:
:+1:

Commit:
Remove description in __openerp__.py

alexis-via committed:
Move ACL from l10n_fr_intrastat_product module to intrastat_product module

alexis-via commented:
Just a quick note to explain that I'm currently working with Luc de Meyer to reorganise the code of the intrastat module to put much more things in common between the different EU countries, starting with FR, BE and NL. So:
- intrastat_product will grow a lot, with all the base objects for the declaration
- we will move the HS code in another module (product_harmonized_system), because companies outside EU need it and there is no point for them to install "intrastat_product" and "intrastat_base". The module intrastat_product will depend on the module product_harmonized_system.

You'll get more updates soon at the end of my code sprint with Luc, which starts today and ends tomorrow.

yvaucher commented:
@alexis-via Good to know it will be improved on use case of 3 different countries this should give a very good base code for other EU countries to join.

And thanks for the changes :+1:

yvaucher commented:
I set this as WIP as @alexis-via and @luc-demeyer are working on it

alexis-via committed:
Large code re-engineering following the Akretion-Noviat code sprint on intrastat

This code has been written both by Luc de Meyer and myself.

alexis-via committed:
Add first draft of code to generate decl lines

gurneyalex commented:
travis failures related to #94 fixed (hopefully) by #96

gurneyalex commented:
My PR fixing the travis and runbot builds on the 8.0 branch were recently merged. Can you rebase your PR?

luc-demeyer committed:
hs_code display_name fix

luc-demeyer commented:
@alexis-via 
alexis,
you have set ondelete='restrict' on the hs_code_id field for product.template and product.category (cf. product_harmonized_system/product.py). I don't think this is needed (and since the fields are properties Odoo doesn't enforce the ondelete hence it's useless anyhow).

luc-demeyer committed:
Update README.rst

luc-demeyer committed:
Update intrastat.py

remove print statements from def _display_name

Commit:
Merge pull request #1 from luc-demeyer/patch-1

hs_code display_name fix

Commit:
Merge pull request #2 from luc-demeyer/patch-2

Update README.rst

Commit:
Merge pull request #3 from luc-demeyer/patch-3

Update intrastat.py

lepistone commented:
Now that OCA/maintainer-quality-tools#214 is merged, travis should finally run the tests for real. The build is green. The lint check finds only two minor issues. :sparkles:

gurneyalex commented:
there are also warnings found on runbot which are easy to fix.

sbidoul commented:
@alexis-via @luc-demeyer is this still Work In Progress?

luc-demeyer commented:
Still work in progress.

damdam-s commented on intrastat_product/account_invoice.py:
```diff
@@ -0,0 +1,88 @@
+
+        if product:
+            product = self.env['product.product'].browse(product)
+            hs_code = product.get_hs_code_recursively()
```
``` python
File ".../account-financial-reporting/intrastat_product/account_invoice.py", line 86, in product_id_change
    hs_code = product.get_hs_code_recursively()
AttributeError: 'product.product' object has no attribute 'get_hs_code_recursively'
```

In fact, it's 'product.template' object which have this method.
So

``` python
hs_code = product.product_tmpl_id.get_hs_code_recursively()
```

should work

@alexis-via : Fixed in PR : https://github.com/akretion/account-financial-reporting/pull/4 on your branch

luc-demeyer replied:
@damdam-s 
I am working on an update which already includes the fix suggested by you as well as a lot of other fixes and improvements.
I have done this work about a month ago but didn't submit a PR yet since there are still some pieces that I needed to complete and I have been busy on other projects.
I expect to finish this work (including the Belgian intrastat localisation on top of it) within the coming two days after which I'll make a PR. If you want to get a copy of what I currently have I can send it to you (but since I am working on it right now, the code is changing every minute and  hence I think it's better that you wait on the upcoming PR).

damdam-s replied:
@luc-demeyer thanks for your reply. I made a new PR just to fix this issue #124 because I needed this fix for a project. I will wait for you PR and test it as soon as you release it. Have a nice day

luc-demeyer commented:
@alexis-via 
cf. https://github.com/akretion/account-financial-reporting/pull/5 with major update of the intrastat framework modules and a first version of the l10n_be_intrastat_product module on top of the new intrastat framework.

luc-demeyer committed:
major update intrastat V3 modules

luc-demeyer committed:
update intrastat modules

Commit:
Merge pull request #6 from Noviat/8.0-update-intrastat-v3

major update intrastat V3 modules

luc-demeyer committed:
add region to intrastat_product

luc-demeyer committed:
add region to intrastat_product

Commit:
Merge pull request #7 from Noviat/8.0-intrastat_product_add_region

add region to intrastat_product

alexis-via committed:
Add product_origin_country_id on declaration/computation lines

Copy incoterms and destination country from SO to invoice when invoicing from SO
We need weight even when supplementary units is used
Small cleanups and enhancements

alexis-via committed:
Add support for accessory costs

Add default values for intrastat transaction on company
Code cleanup

alexis-via committed:
FIX dependency list

More python method call
Update encoding header

alexis-via committed:
If rounded weight is 0, put 1

Take into account the taxes for B2C
Small code cleanup

alexis-via committed:
Remove field exclude_from_intrastat

Re-organise view of intrastat.product.declaration

alexis-via committed:
Add option intrastat_accessory_costs on company

Set more fields as invisible (localisation should put them visible if they need it)
Fix handling of suppl. units when hs_code is empty on invoice line (but set on product)
Small usability enhancements

alexis-via committed:
Include selection of type of invoice in search, for better perf

Isolate domain in a _prepare method, for easier inheritance
WARNING: I changed the default selection of invoice type ; adapt it in your localization m
odule if necessary
Add intrastat_transaction_in_refund for company (not needed for France, but may be needed elsewhere)
Add a log when an invoice line is skipped

alexis-via committed:
Warning -> UserError

alexis-via committed:
Include product code in warning msg on weight

alexis-via committed:
Inspired by the PR https://github.com/akretion/account-financial-reporting/pull/8 of Luc de Meyer

luc-demeyer committed:
intrastat fixes

Commit:
Merge pull request #9 from Noviat/8.0-intrastat-v3-fixes

intrastat fixes

alexis-via committed:
Improve logs and messages

alexis-via committed:
total_amount is a sum of integers, so it should be an integer

Add transport mode in computation tree view

alexis-via committed:
FIX intrastat_country for invoices without src_dest_country_id

alexis-via commented on intrastat_product/models/stock_warehouse.py:
```diff
@@ -0,0 +1,48 @@
+        locations = self.search(
+            [('parent_left', '<=', self.parent_left),
+             ('parent_right', '>=', self.parent_right)])
+        warehouses = self.search(
```
Mmm, that code is wrong ! I'll fix it !

alexis-via replied:
What is very strange is that it wasn't found earlier. I didn't find it earlier because I skip this code in the FR localisation... but @luc-demeyer you don't skip this code in your localisation, don't you ?

alexis-via committed:
FIX wrong model for seach method

alexis-via committed:
Use stock_picking_invoice_link for a better identification of the intrastat region

With this commit, we now support the following scenario: I order to my supplier a quantity of 50 kg and he delivers/invoices 52kg ; odoo will create an additional invoice line of 2kg which is linked to the stock move, but not to any PO line.

alexis-via commented:
@luc-demeyer I made some important changes in my last commit that involve depending on stock_picking_invoice_link. Tell-me what you think. I almost always use the OCA module stock_picking_invoice_link, so it's not a pb for me... but I can understand it may be a problem if you don't want to use this module.

alexis-via committed:
Demo VAT number should be on EU customers

alexis-via committed:
Modularise a piece of code

gurneyalex committed:
add stock-logistics-workflow to dependencies

this is needed to get stock_picking_invoice_link

gurneyalex commented:
@alexis-via I added the dependency on stock-logistics-workflow on the runbot and made a PR on your branch to get this to build correctly in Travis (https://github.com/akretion/account-financial-reporting/pull/11)

Commit:
Merge pull request #11 from gurneyalex/8.0-add-intrastat-v3

add stock-logistics-workflow to dependencies

alexis-via committed:
Add ACL on hs.code to financial manager (I can't do it in product_harmonized_system because it doesn't depend on account)

alexis-via committed:
Handle scenario where an invoice has products with 0 value (samples for example) and shipping costs (accessory costs) with value > 0.

hbrunn committed:
[FIX] this week's ci errors

hbrunn commented:
@alexis-via I provided https://github.com/akretion/account-financial-reporting/pull/12 to have a green travis again

hbrunn committed:
[FIX] wrong variable name

hbrunn committed:
[FIX] another wrong variable name

alexis-via committed:
FIX error message

alexis-via committed:
Remove required=1 because we can have intrastat unit that will never be mapped to an odoo unit (will always be set manually)

Commit:
Merge pull request #12 from hbrunn/8.0-add-intrastat-v3

[FIX] this week's ci errors

alexis-via committed:
Cleaner view inheritance (works both in v8 and v9)

hbrunn commented:
@alexis-via is this ready to review?

alexis-via commented:
@hbrunn yes sure.
It is already used in production by some of our customers.
I know there are some things that @luc-demeyer want to change/update, but we didn't have the time to finish it during the Sorrento code sprint...I hope we will have time to finish during the Odoo Experience sprint.

And by the way, I eventually think that intrastat needs a dedicated OCA repo and that it wasn't such a good idea to propose it to account-financial-reporting...

hbrunn commented:
I think you're right for the dedicated repo. But please lets not tell anyone about it and merge this first. :+1:  here

alexis-via commented:
Mmm, if we decide to have a dedicated repo, we'd better create it now and merge it in the new repo before merging it here I think.

hbrunn commented:
then let's open the new repo. Do we need approval from @OCA/board or can we just push forward?

sbidoul commented:
I'am in favor of a new repo. To follow the procedure, send an email to contribute@odoo-community.org the ask for the creation of the new repo under a PSC to be chosen in the [list](https://odoo-community.org/page/List).

luc-demeyer committed:
update intrastat_base

alexis-via commented:
New repo is here: https://github.com/OCA/intrastat

luc-demeyer committed:
update intrastat_product

luc-demeyer committed:
update intrastat_product

Commit:
Merge pull request #14 from Noviat/oca_sprint_2016

update intrastat_base

alexis-via committed:
Update README.rst

Rename variables

pedrobaeza commented:
Closing as this should be in OCA/instrastat
